File: hackerrank/Domains/Algorithms/DynamicProramming/moderate_equal/equal.cpp
#include <bits/stdc++.h>

using namespace std;

int equal(vector <int> arr) {
    // Complete this function
	int nummoves, move5, move2, move1, mnmoves=-1;
	#the '-4' is a bit hacky. passes test 15
	for (int smallest = -4; smallest <= *std::min_element(arr.begin(), arr.end()); smallest++)
	{
		nummoves = 0;
		for(int k : arr)
		{
			if (k!=smallest)
			{
				move5 = (k-smallest)/5;
                move2 = (k-smallest-move5*5)/2;
                move1 = (k-smallest-move5*5-move2*2);
				nummoves += (move5+move2+move1);
				if (mnmoves >= 0)
                    if (nummoves > mnmoves)
                        break;
			}
		}
		if (mnmoves > nummoves || mnmoves < 0)
            mnmoves = nummoves;
	}
	return mnmoves;
}

int main() {
    int t;
    cin >> t;
    for(int a0 = 0; a0 < t; a0++){
        int n;
        cin >> n;
        vector<int> arr(n);
        for(int arr_i = 0; arr_i < n; arr_i++){
           cin >> arr[arr_i];
        }
        int result = equal(arr);
        cout << result << endl;
    }
    return 0;
}

File: hackerrank/Domains/Algorithms/DynamicProramming/moderate_equal/equal.py
#!/bin/python

import sys

def equal(arr):
    # Complete this function
    #for smallest in range(min(arr),-1,-1):
    mnmoves = -1
    #the '-4' is a bit hacky. passes test 15
    for smallest in range(-4,min(arr)+1):
        nummoves = 0
        for k in arr:
            if k != smallest:
                move5 = (k-smallest)/5
                move2 = (k-smallest-move5*5)/2
                move1 = (k-smallest-move5*5-move2*2)
                nummoves += (move5+move2+move1)
                if mnmoves >= 0:
                    if nummoves > mnmoves:
                        break
        if mnmoves > nummoves or mnmoves < 0:
            mnmoves = nummoves
        print mnmoves
    return mnmoves
    

if __name__ == "__main__":
    
    t = int(raw_input().strip())
    for a0 in xrange(t):
        n = int(raw_input().strip())
        arr = map(int, raw_input().strip().split(' '))
        result = equal(arr)
        print result
    '''
    print equal([2,2,3,7])
    #print equal([2, 5, 5, 5, 5, 5]) #expected 6, got 10
    '''
